File: src/main/java/daykin/rob/dndapi/model/Character.java
package daykin.rob.dndapi.model;

import com.fasterxml.jackson.annotation.JsonAlias;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import lombok.*;

import javax.persistence.*;
import java.util.Set;

@Entity
@Builder
@Getter
@Setter
@Table(name="characters")
@NoArgsConstructor
@AllArgsConstructor
@JsonIgnoreProperties(ignoreUnknown = true)
public class Character {

    @Id
    @JsonAlias("_id")
    String id;
    @Column(unique = true, length = 2047)
    String name;
    int strength;
    int dexterity;
    int constitution;
    int intelligence;
    int wisdom;
    int charisma;
    @JsonAlias("hit_points")
    int hitpoints;
    int proficiencyBonus;
    @JsonAlias("armor_class")
    int armourClass;
    @JsonAlias("challenge_rating")
    int challengeRating;
    String size;
    String type;
    String subtype;
    String alignment;
    String speed;
    @Column(length = 2047)
    String senses;

    @JsonAlias("damage_vulnerabilities")
    @Column(length = 2047)
    String vulnerabilities;

    @JsonAlias("damage_resistances")
    @Column(length = 2047)
    String resistances;

    @JsonAlias("damage_immunities")
    @Column(length = 2047)
    String immunities;

    @JsonAlias("condition_immunities")
    @Column(length = 2047)
    String conditionImmunities;

    @Column(length = 2047)
    String languages;
    int athletics;
    int acrobatics;
    @JsonAlias("sleight_of_hand")
    int sleightOfHand;
    int stealth;
    int Arcana;
    int history;
    int investigation;
    int nature;
    int religion;
    @JsonAlias("animal_handling")
    int animalHandling;
    int insight;
    int medicine;
    int perception;
    int survival;
    int deception;
    int intimidation;
    int performance;
    int persuasion;

    @JsonAlias("strength_save")
    int strengthSave;
    @JsonAlias("dexterity_save")
    int dexteritySave;
    @JsonAlias("constitution_save")
    int constitutionSave;
    @JsonAlias("intelligence_save")
    int intelligenceSave;
    @JsonAlias("wisdom_save")
    int wisdomSave;
    @JsonAlias("charisma_save")
    int charismaSave;
    @JsonAlias("hit_dice")
    String hitDice;

    @OneToMany(cascade=CascadeType.ALL)
    @JoinColumn(name = "character_id")
    Set<Action> actions;

    @OneToMany(cascade=CascadeType.ALL)
    @JoinColumn(name = "character_id")
    Set<SpecialAbility> legendaryActions;


    @OneToMany(cascade=CascadeType.ALL)
    @JoinColumn(name = "character_id")
    Set<SpecialAbility> specialAbilities;

    @Override
    public boolean equals(Object object){
        if(object == null || object.getClass() != Character.class){
            return false;
        }
        Character character = (Character) object;
        return character.getStrength() == strength
                && character.getDexterity() == dexterity
                && character.getConstitution() == constitution
                && character.getIntelligence() == intelligence
                && character.getWisdom() == wisdom
                && character.getCharisma() == charisma
                && character.getHitpoints() == hitpoints
                && character.getSpeed() == speed
                && character.getProficiencyBonus() == proficiencyBonus;
    }
}

File: reactui/react-api/src/components/character.js
// src/components/skills.js

import React, {Component}  from 'react'
import SavingThrows from './savingthrows'
import Skills from './skills'
import Actions from './action'

class Character extends Component {

  state = {
    character:[]
  }

  constructor(props){
    super(props)
    this.state.character = props.character
  }

  render(){
    var character = this.props.character
    return (
      <center>
        <div>
          <h1>{character.name}</h1>
          <div>
            <strong>Armour Class</strong> {character.armourClass}<br/>
            <strong>Hit Points</strong> {character.hitpoints} ( {character.hitDice})<br/>
            <strong>Speed</strong> {character.speed}<br/>
          </div>
          <div>
            <table className="table table-bordered table-condensed">
              <thead>
                <tr>
                  <th><strong>STR</strong></th>
                  <th><strong>DEX</strong></th>
                  <th><strong>CON</strong></th>
                  <th><strong>INT</strong></th>
                  <th><strong>WIS</strong></th>
                  <th><strong>CHA</strong></th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td>{character.strength}</td>
                  <td>{character.dexterity}</td>
                  <td>{character.constitution}</td>
                  <td>{character.intelligence}</td>
                  <td>{character.wisdom}</td>
                  <td>{character.charisma}</td>
                </tr>
              </tbody>
            </table>
          </div>
          {character.name ? <SavingThrows character={character}/>: null}
          {character.name ? <Skills character={character}/>: null}
          {character.resistances && character.resistances != "" ? <strong>Damage Resistances </strong>:null} {character.resistances} {character.resistances !== "" ? <br/>:null}
          {character.immunities && character.immunities != "" ? <strong>Damage Immunities </strong>:null} {character.immunities} {character.immunities !== "" ? <br/>:null}
          {character.conditionImmunities && character.conditionImmunities != "" ? <strong>Condition Immunities </strong>:null} {character.conditionImmunities} {character.conditionImmunities !== "" ? <br/>:null}
          {character.senses && character.senses !== "" ? <strong>Senses </strong>:null} {character.senses} {character.senses !== "" ? <br/>:null}
          {character.languages && character.languages !== "" ? <strong>Languages </strong>:null} {character.languages} {character.languages !== "" ? <br/>:null}
          <strong>Challenge</strong> {character.challengeRating}
          {character.specialAbilities && character.specialAbilities !== [] ? <h2>Special Abilities</h2>: null}
          {character.specialAbilities && character.specialAbilities !== [] ? <Actions actions={character.specialAbilities}/>: null}
          {character.actions && character.actions !== [] ? <h2>Actions</h2>: null}
          {character.actions? <Actions actions={character.actions}/>: null}
        </div>
      </center>
    )
  }
};

export default Character
